### PR TITLE
feat(https): optional HTTPS proxy

### PR DESCRIPTION
**Pull request by mdibaiee**

### CODE REVIEW

File: app/Main.hs [hunk 1]
```diff
@@ -20,6 +23,9 @@ defaultSettings = Settings { bindAddress = "0.0.0.0"
                            , bufferSize = 2^11
                            , authentication = ""
                            , realm = ""
+                           , https = False
+                           , cert = "cert.pem"
```
Comment (amir-sabbaghi) [inline app/Main.hs:27]: Don't use default filenames for cert & key. It's only useful for testing.

File: app/Main.hs [hunk 2]
```diff
@@ -20,6 +23,9 @@ defaultSettings = Settings { bindAddress = "0.0.0.0"
                            , bufferSize = 2^11
                            , authentication = ""
                            , realm = ""
+                           , https = False
```
Comment (amir-sabbaghi) [inline app/Main.hs:26]: I think `https` should have a type with 2 constructors: `HTTP` and `HTTPS FilePath FilePath`

Comment (mdibaiee) [reply app/Main.hs:26]: I was thinking about having HTTP and HTTPS at the same time using a single process.

What do you think about this?

``` haskell
http = Just (HTTP "8080") | Nothing
https = Just (HTTPS "8080" "cert.pem" "key.pem") | Nothing
```

Comment (amir-sabbaghi) [reply app/Main.hs:26]: Sounds good.

**Comment by mdibaiee** [comment]
We're going to use `tls` instead of `HsOpenSSL` because HsOpenSSL doesn't support ALPN.

See https://github.com/amir-sabbaghi/proxy/pull/5

**Pull request closed without merging**